extension/intents/timer/timer.js

```diff
@@ -0,0 +1,183 @@
+/* globals catcher */
+
+import * as intentRunner from "../../background/intentRunner.js";
+
+class TimerController {
+  constructor() {
+    this.activeTimer = null;
+  }
+
+  getActiveTimer() {
+    return this.activeTimer;
+  }
+
+  closeActiveTimer() {
+    this.activeTimer.close();
+    this.activeTimer = null;
+  }
+
+  setActiveTimer(totalInMS) {
+    this.activeTimer = new Timer(totalInMS);
+    this.activeTimer.start();
+  }
+}
+
+class Timer {
+  constructor(totalInMS) {
+    this.totalInMS = totalInMS;
+    this.startTimestamp = undefined;
+    this.remainingInMS = undefined;
+    this.paused = undefined;
+  }
+
+  close() {
+    this.startTimestamp = undefined;
+    this.totalInMS = undefined;
+    this.remainingInMS = undefined;
+    if (this.timeoutId !== undefined) clearTimeout(this.timeoutId);
+  }
+
+  pause() {
+    this.paused = true;
+    this.remainingInMS = this.remainingMs();
+    if (this.timeoutId !== undefined) clearTimeout(this.timeoutId);
+  }
+
+  async unpause() {
+    this.paused = false;
+    return this.start(this.remainingInMS);
+  }
+
+  remainingMs() {
+    return this.totalInMS - (new Date().getTime() - this.startTimestamp);
+  }
+
+  reset() {
+    if (this.timeoutId !== undefined) clearTimeout(this.timeoutId);
+
+    return this.start(this.totalInMS);
+  }
+
+  async start(duration = this.totalInMS) {
+    this.startTimestamp = new Date().getTime();
+    this.remainingInMS = duration;
+
+    this.timeoutId = setTimeout(() => this.openPopup(), duration);
```

Comment:
There's a few blank lines and I prefer keeping braces around all conditionals, but not blockers...